In [5]:
import pandas as pd 
import os 

In [6]:
# carga de datos en csv

data = pd.read_csv('Ordenes_producto_C1_M2.csv', sep = ';', encoding = 'latin-1')
data.head()

orden_id order_item_id producto_id vendedor_id fecha_envio_limite  \
0  107500PO59A             A        PO59      VE5389   07/04/2018 18:12   
1   37493PS22B             B        PS22      VE1558   20/10/2017 09:07   
2   28050PK20B             B        PK20      VE9159   17/08/2017 08:15   
3   52187PA10A             A        PA10      VE3159   23/09/2017 23:27   
4   84639PR12A             A        PR12      VE5090   07/01/2018 11:50   

   precio  valor_flete  codigo_postal_vendedor ciudad_vendedor  \
0  271.86        30.72                   52435         Mallama   
1  115.73         4.68                   52203           Colon   
2  432.99        82.70                   66001         Pereira   
3  108.38        35.39                   52435         Mallama   
4   51.50        11.10                   73001          Ibague   

  departamento_vendedor nombre_categoria_producto  longitud_nombre_producto  \
0                Nariño    Productos ecoamigables                         6   
1                Nariño                Carnicería                        10   
2             Risaralda                  Deportes                        25   
3                Nariño         Electrodomésticos                        10   
4                Tolima         Frutas y verduras                        23   

   longitud_descripcion_producto  cantidad_fotos_producto  peso_g_producto  \
0                              7                       27             2486   
1                             31                       20              256   
2                              5                        4             5270   
3                              1                        6              734   
4                             16                       35              884   

   longitud_cm_producto  altura_cm_producto  ancho_cm_producto  
0                    17                  11                 14  
1                    43                   2                 21  
2                     9                  27                 29  
3                    46                  48                 22  
4                    45                  26                 18

In [7]:
# Revisamos el tipo de cada una de las variables 
data.dtypes

orden_id                          object
order_item_id                     object
producto_id                       object
vendedor_id                       object
fecha_envio_limite                object
precio                           float64
valor_flete                      float64
codigo_postal_vendedor             int64
ciudad_vendedor                   object
departamento_vendedor             object
nombre_categoria_producto         object
longitud_nombre_producto           int64
longitud_descripcion_producto      int64
cantidad_fotos_producto            int64
peso_g_producto                    int64
longitud_cm_producto               int64
altura_cm_producto                 int64
ancho_cm_producto                  int64
dtype: object

## Tablas de frecuencia 

Para la solución del caso de uso es necesario identificar una posible hipótesis para validar a través de las herramientas que se revisarán en las siguientes líneas.

La tabla de frecuencia es una herramienta que permite ordenar los datos que se presentan numéricamente las características de la distribución de un conjunto de datos o muestra. 

**Hipótesis:** la categoría de celulares es la que más productos vende, se quiere validar ya que se desea lanzar una campaña de descuento para traer más clientes a la empresa, sin embargo, se debe conocer cuál va a ser la categoría en la cual se invertirá más publicidad. 

In [8]:
data.columns

Index(['orden_id', 'order_item_id', 'producto_id', 'vendedor_id',
       'fecha_envio_limite', 'precio', 'valor_flete', 'codigo_postal_vendedor',
       'ciudad_vendedor', 'departamento_vendedor', 'nombre_categoria_producto',
       'longitud_nombre_producto', 'longitud_descripcion_producto',
       'cantidad_fotos_producto', 'peso_g_producto', 'longitud_cm_producto',
       'altura_cm_producto', 'ancho_cm_producto'],
      dtype='object')

In [18]:
## Tabla de frecuencia univariada 
# Se va a realizar una tabla para analizar las categorías que se venden en el comercio. 
# Frecuencia absoluta

frec_data = data['nombre_categoria_producto'].value_counts().reset_index()
frec_data.head()

index  nombre_categoria_producto
0         Tecnología                        543
1           Deportes                        537
2  Frutas y verduras                        532
3         Carnicería                        523
4    Ropa de adultos                        521

In [19]:
# Frecuencia relativa 
frec_data['porcentaje_nombre_categoria_producto'] = frec_data['nombre_categoria_producto'].apply(lambda x: x/frec_data['nombre_categoria_producto'].sum())
frec_data.head(5)

index  nombre_categoria_producto  \
0         Tecnología                        543   
1           Deportes                        537   
2  Frutas y verduras                        532   
3         Carnicería                        523   
4    Ropa de adultos                        521   

   porcentaje_nombre_categoria_producto  
0                              0.053582  
1                              0.052990  
2                              0.052497  
3                              0.051608  
4                              0.051411

In [20]:
# Tabla de frecuencia absoluta y relativa acumulada con comsum()
## Frecuencia absoluta acumulada 
frec_data['acumulada_nombre_categoria_producto']=frec_data['nombre_categoria_producto'].cumsum()

## Frecuencia relativa acumulada 
frec_data['acumulada_porcentaje_nombre_categoria_producto']=frec_data['porcentaje_nombre_categoria_producto'].cumsum()
frec_data.head()

index  nombre_categoria_producto  \
0         Tecnología                        543   
1           Deportes                        537   
2  Frutas y verduras                        532   
3         Carnicería                        523   
4    Ropa de adultos                        521   

   porcentaje_nombre_categoria_producto  acumulada_nombre_categoria_producto  \
0                              0.053582                                  543   
1                              0.052990                                 1080   
2                              0.052497                                 1612   
3                              0.051608                                 2135   
4                              0.051411                                 2656   

   acumulada_porcentaje_nombre_categoria_producto  
0                                        0.053582  
1                                        0.106572  
2                                        0.159068  
3                                        0.210677  
4                                        0.262088

In [21]:
frec_data

index  nombre_categoria_producto  \
0               Tecnología                        543   
1                 Deportes                        537   
2        Frutas y verduras                        532   
3               Carnicería                        523   
4          Ropa de adultos                        521   
5               Ferretería                        519   
6                 Juguetes                        519   
7            Ropa infantil                        512   
8        Electrodomésticos                        510   
9               Dormitorio                        509   
10  Productos ecoamigables                        501   
11                 Lácteos                        501   
12                  Libros                        499   
13         Tarjetas regalo                        497   
14                Mascotas                        496   
15                   Salud                        491   
16               Celulares                        488   
17                 Muebles                        486   
18                   Licor                        475   
19                  Bebés                         475   

    porcentaje_nombre_categoria_producto  acumulada_nombre_categoria_producto  \
0                               0.053582                                  543   
1                               0.052990                                 1080   
2                               0.052497                                 1612   
3                               0.051608                                 2135   
4                               0.051411                                 2656   
5                               0.051214                                 3175   
6                               0.051214                                 3694   
7                               0.050523                                 4206   
8                               0.050326                                 4716   
9                               0.050227                                 5225   
10                              0.049438                                 5726   
11                              0.049438                                 6227   
12                              0.049240                                 6726   
13                              0.049043                                 7223   
14                              0.048944                                 7719   
15                              0.048451                                 8210   
16                              0.048155                                 8698   
17                              0.047957                                 9184   
18                              0.046872                                 9659   
19                              0.046872                                10134   

    acumulada_porcentaje_nombre_categoria_producto  
0                                         0.053582  
1                                         0.106572  
2                                         0.159068  
3                                         0.210677  
4                                         0.262088  
5                                         0.313302  
6                                         0.364515  
7                                         0.415038  
8                                         0.465364  
9                                         0.515591  
10                                        0.565029  
11                                        0.614466  
12                                        0.663706  
13                                        0.712749  
14                                        0.761693  
15                                        0.810144  
16                                        0.858299  
17                                        0.906256  
18                                        0.953128  
19                                        1.000000

*Análisis**: La tabla de frecuencias nos muestra la siguiente información:
1. Frecuencia absoluta: número de veces que aparece un valor.
2. Frecuencia relativa: el resultado de dividir la frecuencia absoluta de un determinado valor por el número total de datos. 
3. Frecuencia acumulada: la suma de las frecuencias absolutas de todos los valores iguales o menores al valor considerado.
4. Frecuencia acumulada relativa: el resultado de dividir la frecuencia acumulada por el número de toda la información. 

In [22]:
## Tablas de frecuencia bivariada
# Tabla que muestra la frecuencia por ciudad de venta y categoría del producto
pd.crosstab(index=data['nombre_categoria_producto'], columns=data['ciudad_vendedor'])

ciudad_vendedor            Abejorral  Abrego  Abriaqui  Acacias  Acandi  \
nombre_categoria_producto                                                 
Bebés                              9      23         1       20       0   
Carnicería                         8      16         1       15       0   
Celulares                          7      19         0       17       0   
Deportes                          10      19         0       24       1   
Dormitorio                         8       9         2       19       0   
Electrodomésticos                  5      27         1       17       1   
Ferretería                         5      14         0       18       0   
Frutas y verduras                  6      20         0       16       2   
Juguetes                           9      18         0       19       2   
Libros                             7      18         1       20       2   
Licor                              6      16         0       14       0   
Lácteos                            6      12         1       19       0   
Mascotas                           9      12         0       13       1   
Muebles                            3      19         1       19       0   
Productos ecoamigables             8      15         0       20       1   
Ropa de adultos                    4      20         1       20       1   
Ropa infantil                     12      22         2       15       0   
Salud                              5      16         1       11       0   
Tarjetas regalo                    9      20         1       21       1   
Tecnología                         8      20         2       20       1   

ciudad_vendedor            Acevedo  Achi  Agua de dios  Arauca  Arboleda  ...  \
nombre_categoria_producto                                                 ...   
Bebés                            0     9             4       3         6  ...   
Carnicería                       0     5             7      10         7  ...   
Celulares                        1    11             9       4         9  ...   
Deportes                         0     4             3       5         5  ...   
Dormitorio                       0     9             3      11         9  ...   
Electrodomésticos                0     5             6       3         4  ...   
Ferretería                       0     6             7       9         6  ...   
Frutas y verduras                0     6             5       8        12  ...   
Juguetes                         0    11             3       6        11  ...   
Libros                           0     9             9       5         8  ...   
Licor                            1    10             4       9         2  ...   
Lácteos                          3    10             8       7         6  ...   
Mascotas                         0     7             7       4         7  ...   
Muebles                          2     5             3       1        11  ...   
Productos ecoamigables           2     9             4       6        14  ...   
Ropa de adultos                  0    11             3       3        10  ...   
Ropa infantil                    2    11             3      11        11  ...   
Salud                            0     6             4      10         3  ...   
Tarjetas regalo                  0     9             5      13         4  ...   
Tecnología                       1     6             4       3         8  ...   

ciudad_vendedor            San andres  San jose del guaviare  \
nombre_categoria_producto                                      
Bebés                               2                      6   
Carnicería                          1                      6   
Celulares                           3                      3   
Deportes                            3                      6   
Dormitorio                          2                      6   
Electrodomésticos                   2                      1   
Ferretería                          3                 

**Análisis**: Es posible en la tabla ver la cantidad de compras que tiene las diferentes categorías, la que más tiene es *Tecnología* con 532 ventas en total. 

## Tendencia central 

La tendencia central significa medir el centro o distribución de ubicación de valores de un conjuto de datos. Da una idea del valor promedio de las variables en el conjunto de datos y también una indicación de cuán ampliamente se distribuyen los valore en el conjunto de datos.

1. Media
2. Mediana
3. Moda

**Hipótesis**: Los precios en el comercio son mayores a 500, ya que el estudio arrojó que los clientes más recurrentes compran mayormente productos más ecnonómicos (debajo de 500), por tanto, es necesario validar si en promedio la compañía tiene precios más competitivos.

In [23]:
## Media o Promedio 
# Promedio de precios
round(data['precio'].mean(),2)

119.79

In [24]:
## Mediana
# Mediana de precio, valor central de la distribución de los datos.
round(data['precio'].median(),2)

83.12

In [25]:
## Moda
# Valor más común de los precios
data['precio'].mode()

0     3.92
1    11.39
2    11.79
3    15.80
4    16.42
5    25.45
6    34.43
7    36.96
8    56.79
Name: precio, dtype: float64

**Análisis**: Es posible analizar que el promedio es inferior a 500 ubicándose en 120, la mediana que es el dato central se ubica en 83.12 y existen varios valores comunes y todos debajo de 60 lo que nos muestra que la mayor cantidad de precios son bajos y que permiten ser más atractivos para el consumidor final que compra productos mayormente debajo de 500.

## Medidas de localización 

Los *cuantiles* y *percentiles* son puntos de corte que dividen el rango de una distribución de probabilidad en intervalos continuos son probabilidades iguales o que dividen a las observaciones en una muestra de la misma forma.

Así se puede representar un porcentaje de datos a través de los cuantiles para indetificar cómo están dividos los datos. 

**Hipótesis**: la mitat (50%) de los productos pesa más de 500 gramos, ya que la empresa de mensajería quiere realizar un descuento si los productos pesan menos de 500 gramos, así la compañía podría bajar los precios de envío de la mitad de los producto que se venden.

In [27]:
## Cuantiles 
# Calcular el 20%, 50% y 70% del peso de los productos. 
data['peso_g_producto'].quantile([0.20,0.50,0.70])

0.2     470.0
0.5    1458.5
0.7    2535.1
Name: peso_g_producto, dtype: float64

**Análisis**: El percentil 50 tiene un valor de 1458.5 lo cual nos indica que la mitad de los productos tiene un peso de más alto que 500 gramos, por lo que se rechaza la hipótesis y nos indica que no se puede realizar el descuento sobre el vólumen del producto, sin embargo deja abierta la puerta para encontrar qué cantidad de producto pueden contar con el descuento. 

## Medidas de dispersión 

Se utilizan las estadísticas de dispersión para dar un número único que describa qué tan compacto o disperso es un conjunto de observaciones. Estas se cumplementan de las medidas mencionadas anteriormente. 

Estas medidas entre _más grandes_, representan más dispersión y entre más pequeñas más concentradas están las observaciones (es importante tener en cuenta la escala de la variable para analizar la dispersión).

1. Desviación estándar
2. Rango intercuartílico
3. Rango 
4.  Coeficiente de variación

**Hipótesis**: La variación de las longitudes de los datos es mayor a 20cm, ya que las empresas de mensajería que tienen cubículos de transporte de mensajería individual de producto de longitud de 20 cm son más económicas, y así se podría cambiar estas empresas y ahorrar costos de envío . 

In [29]:
## Desviación estándar 
# Encontrar la desviación estándar de la longitud de los productos de la muestra
# Represental qué tan lejos de la media se encuentran las observaciones 
round(data['longitud_cm_producto'].std(),2)

15.29

In [30]:
## Rango intercuartílico 
# Encontra el rango intercuartílico de la longitud de los productos
# es la diferencia entre el tercer y primer cuartil. 
round(data['longitud_cm_producto'].quantile(0.75)-data['longitud_cm_producto'].quantile(0.25),2)

22.0

In [31]:
# Rango 
# Encontrar el rango de la longitud de los productos de la muestra 
# Representa la diferencia que existe entre el producto más y el más pequeño
data['longitud_cm_producto'].max()-data['longitud_cm_producto'].min()

91

In [33]:
## Coeficiente de variación 
# Encontrar el coeficiente de variación de la longitud de los productos. 
# Es la división de la desviación estándar sobre la media
round(data['longitud_cm_producto'].std()/data['longitud_cm_producto'].mean(),3)

0.509

**Análisis**: a pesar de que la desviación estándar y el rango intercuartílico son menores a 25, se puede alcanzar con el rango que la variación entre el máximo y el mínimo es de 91 cm, por tanto se podría migrar alredador del 50% de los productos de mensajería con el cubículo de 20 cm y darle al cliente un descuento en el envío. 

### Bonus

En la librería de pandas es posible encontrar una función que agrega varias estadísticas que permiten analizar rápidamente la distribución de los datos numéricos de un data frame. 

In [34]:
data.describe()

precio   valor_flete  codigo_postal_vendedor  \
count  10134.000000  10134.000000            10134.000000   
mean     119.792347     20.053246            49991.684626   
std      119.340785     20.106163            22463.895340   
min        0.010000      0.000000             5001.000000   
25%       34.155000      5.760000            50001.000000   
50%       83.120000     13.980000            52323.000000   
75%      165.515000     27.455000            54003.000000   
max     1262.940000    183.150000            99001.000000   

       longitud_nombre_producto  longitud_descripcion_producto  \
count              10134.000000                   10134.000000   
mean                  20.074798                      19.845372   
std                   11.721956                      11.854161   
min                    0.000000                       0.000000   
25%                   10.000000                      10.000000   
50%                   20.000000                      20.000000   
75%                   30.000000                      30.000000   
max                   40.000000                      40.000000   

       cantidad_fotos_producto  peso_g_producto  longitud_cm_producto  \
count             10134.000000     10134.000000          10134.000000   
mean                 20.110026      2106.109236             30.071936   
std                  11.802456      2093.734472             15.292866   
min                   0.000000         0.000000              0.000000   
25%                  10.000000       607.000000             19.000000   
50%                  20.000000      1458.500000             30.000000   
75%                  30.000000      2941.000000             41.000000   
max                  40.000000     20661.000000             91.000000   

       altura_cm_producto  ancho_cm_producto  
count        10134.000000       10134.000000  
mean            17.915433          22.905861  
std             11.360360          11.151559  
min              0.000000           0.000000  
25%              9.000000          15.000000  
50%             17.000000          23.000000  
75%             26.000000          30.000000  
max             64.000000          72.000000